File: DesignPatterns/01_GOF23/DP_examples/CSharp/01_Creational/Abstract Factory/AbstractProduct.cs
namespace Creational;

public interface IColor { void Fill(); }
public class Red : IColor {
    void IColor.Fill() => Console.WriteLine("Fill Red");
}
public class Green : IColor {
    void IColor.Fill() => Console.WriteLine("Fill Green");
}
public class Blue : IColor {
    void IColor.Fill() => Console.WriteLine("Fill Blue");
}
internal class UnknownColor : IColor {
    internal static readonly UnknownColor Default = new UnknownColor();
    void IColor.Fill() { }
}
public interface IShape { void Draw(); }
public class Circle : IShape {
    void IShape.Draw() => Console.WriteLine("Draw a Circle");
}
public class Square : IShape {
    void IShape.Draw() => Console.WriteLine("Draw a Square");
}
public class Rectangle : IShape {
    void IShape.Draw() => Console.WriteLine("Draw a Rectangle");
}
internal class UnknownShape : IShape {
    internal static readonly UnknownShape Default = new UnknownShape();
    void IShape.Draw() { }
}

File: DesignPatterns/01_GOF23/DP_examples/Python/01_Creational/Abstract Factory/AbstractProduct.py
"""
AbstractProduct.py
产品接口和具体实现
"""

from abc import ABC, abstractmethod

# 产品接口
class IColor(ABC):
    @abstractmethod
    def Fill(self):
        pass

class IShape(ABC):
    @abstractmethod
    def Draw(self):
        pass

# 具体产品 - 颜色
class Red(IColor):
    def Fill(self):
        print("Fill Red")
class Green(IColor):
    def Fill(self):
        print("Fill Green")
class Blue(IColor):
    def Fill(self):
        print("Fill Blue")
class UnknownColor(IColor):
    def Fill(self):
        pass

# 具体产品 - 形状
class Circle(IShape):
    def Draw(self):
        print("Draw a Circle")
class Square(IShape):
    def Draw(self):
        print("Draw a Square")
class Rectangle(IShape):
    def Draw(self):
        print("Draw a Rectangle")
class UnknownShape(IShape):
    def Draw(self):
        pass
